File: src/skillguard/core/hasher.py
"""SHA256 hashing utilities for skill files and content."""

from __future__ import annotations

import hashlib
from pathlib import Path


def hash_content(content: str) -> str:
    """Compute SHA256 hash of text content."""
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


def hash_file(path: str | Path) -> str:
    """Compute SHA256 hash of a file by reading in chunks."""
    h = hashlib.sha256()
    with open(path, "rb") as f:
        while chunk := f.read(8192):
            h.update(chunk)
    return h.hexdigest()


def hash_skill(files: list[tuple[str, str]]) -> str:
    """Compute a composite hash for an entire skill package.

    Args:
        files: List of (relative_path, file_sha256) tuples, will be sorted.

    Returns:
        SHA256 hash of the concatenated sorted path:hash pairs.
    """
    parts = sorted(f"{path}:{sha}" for path, sha in files)
    combined = "\n".join(parts)
    return hashlib.sha256(combined.encode("utf-8")).hexdigest()

File: go-skillguard/internal/core/hasher.go
package core

import (
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
)

// HashContent computes the SHA256 hash of a string.
func HashContent(content string) string {
	h := sha256.Sum256([]byte(content))
	return fmt.Sprintf("%x", h)
}

// HashFile computes the SHA256 hash of a file by reading in chunks.
func HashFile(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := sha256.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}
	return fmt.Sprintf("%x", h.Sum(nil)), nil
}

// HashSkill computes a composite hash for an entire skill package.
// Takes a slice of (relativePath, fileSHA256) pairs.
func HashSkill(files []FileHash) string {
	parts := make([]string, len(files))
	for i, f := range files {
		parts[i] = f.Path + ":" + f.SHA256
	}
	sort.Strings(parts)
	combined := strings.Join(parts, "\n")
	h := sha256.Sum256([]byte(combined))
	return fmt.Sprintf("%x", h)
}

// FileHash is a path:hash pair for composite hashing.
type FileHash struct {
	Path   string
	SHA256 string
}
